src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,124 @@
+package htsjdk.samtools.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncWriterPool<A> implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed;
+    private final ArrayList<PooledWriter<A>> writers;
+
+    public AsyncWriterPool(final int threads) {
+        assert threads >= 1; // TODO: throw error
+
+        this.poolClosed = new AtomicBoolean(false);
+        // TODO: make this structure threadsafe
+        this.writers = new ArrayList<>();
+
+        executor = Executors.newWorkStealingPool(threads);
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (PooledWriter<A> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+    public static class PooledWriter<B> implements WrappedWriter<B> {
+        private final WrappedWriter<B> writer;
+        private final BlockingQueue<B> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+        public final AsyncWriterPool<B> outer;
+
+        private final int buffsize;
+        private Future<Integer> currentTask;
+
+
+        public PooledWriter(final AsyncWriterPool<B> outer, final WrappedWriter<B> writer, final BlockingQueue<B> queue, final int buffSize) {
+            assert buffSize >= 1; // TODO: Add error message
+            this.writer = writer;
+            this.queue = queue;
+            this.buffsize = buffSize;
+            this.outer = outer;
+            this.writing = new AtomicBoolean(false);
+            this.isClosed = new AtomicBoolean(false);
+            this.currentTask = null;
+
+            // Add self to writer
+            this.outer.writers.add(this);
+        }
+
+        /**
+         * Will throw any error thrown in task
+         */
+        private void checkAndRethrow() {
+            if (!(this.currentTask == null)) {
+                try {
+                    // NB: ignoring output of number of records previously written
+                    this.currentTask.get();
+                } catch (CancellationException | InterruptedException | ExecutionException e) {
+                    this.isClosed.set(true);
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        @Override
+        public void write(final B item) {
+            if (this.isClosed.get()) throw new RuntimeIOException("Attempt to add record to closed writer.");
```

Comment:
Does `isClosed` need to be an atomic boolean?  I think not.  I.e. if we make it clear that `PooledWriter`s are _not_ inherently thread safe and you should synchronize if multiple threads want to call `write()` or `close()` then this can be a simple `boolean` and will perform better.